libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/AlternativeRouteProgressDataProvider.kt

```diff
@@ -0,0 +1,51 @@
+package com.mapbox.navigation.core.routealternatives
+
+import com.mapbox.navigation.base.route.NavigationRoute
+import com.mapbox.navigation.base.utils.DecodeUtils.stepsGeometryToPoints
+import com.mapbox.navigation.core.RouteProgressData
+
+internal object AlternativeRouteProgressDataProvider {
+
+    fun getRouteProgressData(
+        primaryRouteProgressData: RouteProgressData,
+        alternativeMetadata: AlternativeRouteMetadata
+    ): RouteProgressData {
+        val legIndex: Int
+        val routeGeometryIndex: Int
+        val legGeometryIndex: Int?
+        val primaryFork = alternativeMetadata.forkIntersectionOfPrimaryRoute
+        val alternativeFork = alternativeMetadata.forkIntersectionOfAlternativeRoute
+        val isForkPointAheadOnPrimaryRoute =
+            primaryRouteProgressData.routeGeometryIndex < primaryFork.geometryIndexInRoute
+        if (isForkPointAheadOnPrimaryRoute) {
+            val legIndexDiff = primaryFork.legIndex - alternativeFork.legIndex
+            legIndex = primaryRouteProgressData.legIndex - legIndexDiff
+            val routeGeometryIndexDiff =
+                primaryFork.geometryIndexInRoute - alternativeFork.geometryIndexInRoute
+            routeGeometryIndex =
+                primaryRouteProgressData.routeGeometryIndex - routeGeometryIndexDiff
+            legGeometryIndex = routeGeometryIndex -
+                prevLegsGeometryIndicesCount(alternativeMetadata.navigationRoute, legIndex)
+        } else {
+            legIndex = alternativeFork.legIndex
+            routeGeometryIndex = alternativeFork.geometryIndexInRoute
+            legGeometryIndex = alternativeFork.geometryIndexInLeg
+        }
+        return RouteProgressData(legIndex, routeGeometryIndex, legGeometryIndex)
+    }
+
+    private fun prevLegsGeometryIndicesCount(route: NavigationRoute, currentLegIndex: Int): Int {
+        var result = 0
+        val stepsGeometries = route.directionsRoute.stepsGeometryToPoints()
```

Comment:
```suggestion
        val stepsGeometries: List<List<List<Point>>> by lazy {
            route.directionsRoute.stepsGeometryToPoints()
        }
```
Let's not deserialize the points if not needed (so if the fork is on the current leg).